elasticsearch/class-elasticsearch.php

```diff
@@ -49,6 +54,144 @@ protected function load_commands() {
 		}
 	}
 
+	/**
+	 * Return singleton instance of class
+	 *
+	 * @return object
+	 */
+	public static function factory() {
```

Comment:
Why a Singleton pattern (and the constructor isn't private anyway, so this isn't forcing anything)?

How will you mock this class for unit tests if it's a Singleton?
What benefit does it give?

Reply:
In terms of its use in the `Health_Command` class, instantiating `Elasticsearch` outside of the command class, and injecting it as a dependency to the constructor (and saving that to a property of `Health_Command` class`) would allow `Elasticsearch` to be mocked, and to follow best (SOLID) practices.

The `WP_CLI::add_command(…)` can accept an instantiated object, so you can construct and inject before handing over to WP-CLI. So instead of [this line](https://github.com/Automattic/vip-go-mu-plugins/blob/add/es-validation-taxonomy/elasticsearch/class-elasticsearch.php#L53), you could do:

```php
$hc = new \Health_Command( new Elasticsearch() );
WP_CLI::add_command 'vip-es health, $hc );
```

That way, instead of having to instantiate a new Elasticsearch each time (via the factory method), you can instantiate it once, and re-use the service object to do the verifications.

Reply:
> Why a Singleton pattern (and the constructor isn't private anyway, so this isn't forcing anything)?

Do we really need to have multiple instances of this class?
Also this class has no constructor at all; we may either add a `private` one, or a `public` one which returns the singleton instance.
Singleton was used for the sake of this PR, but assuming that the class will be used by other modules in the future, we may convert to a normal class.
@nickdaugherty thoughts?

> How will you mock this class for unit tests if it's a Singleton?

There are ways to mock the class for unit tests, but perhaps it makes things too complex.

Reply:
Thanks @GaryJones for the input, much appreciated! Also please feel free to contribute to the PR if you'd like to write some code!

Reply:
Moreover: I used singleton but probably I should have used `static` methods for `validate_X` functions.
Ideally, as Nick noted later on, we could also move those methods into their own class.
I will work at refactoring this code to be more consistent and make more sense.

Reply:
> Do we really need to have multiple instances of this class?

Likely not, but then, controlling the number of instantiations AND doing the logic that it contains, is two tasks, and therefore goes against the S in SOLID.

> I will work at refactoring this code to be more consistent and make more sense.

👍